02-peer-protocol.md

```diff
@@ -580,31 +598,51 @@ The receiving node:
   - if `fee_satoshis` is equal to its previously sent `fee_satoshis`:
     - SHOULD sign and broadcast the final closing transaction.
     - MAY close the connection.
-  - otherwise, if `fee_satoshis` is greater than
-the base fee of the final commitment transaction as calculated in
-[BOLT #3](03-transactions.md#fee-calculation):
-    - MUST fail the connection.
-  - if `fee_satoshis` is not strictly
-between its last-sent `fee_satoshis` and its previously-received
-`fee_satoshis`, UNLESS it has since reconnected:
+  - if `fee_satoshis` matches its previously sent `fee_range`:
+    - SHOULD use `fee_satoshis` to sign and broadcast the final closing transaction
+    - SHOULD reply with a `closing_signed` with the same `fee_satoshis` value if it is different from its previously sent `fee_satoshis`
+    - MAY close the connection.
+  - if the message contains a `fee_range`:
+    - if there is no overlap between that and its own `fee_range`:
+      - SHOULD fail the connection
+      - MUST fail the channel if it doesn't receive a satisfying `fee_range` after a reasonable amount of time
```

Comment:
Does this imply that you are allowed to send multiple `closing_signed` messages without "waiting your turn" if you send bad fee ranges? If so, this interacts with musig2+taproot channels

Reply:
Yes, this is not a turn-based protocol, the initiator may send another `closing_signed` with a different fee range.

Reply:
Just to clarify, legacy coop close _is_ turn-based, but fee_range isn't? 

What is the rationale to make this not turn-based? Shouldn't the recipient of a fee_range ensure that the sent fee_range has overlap, or else don't send one?

Reply:
I haven't posted my set of comments there yet but you don't need to use nonces in shutdown -- your point still remains the same as you'd be using up the latest nonces you have. since there's no acknowledgement in `closing_signed`, you either need:
- turn-based coop closing always
- some acknowledgement message
- deterministic nonce points only for the coop closing domain (send these in shutdown, don't use for regular commitments)
- pre-share a bunch of nonces in shutdown (impossible to guess how many nonces is enough without setting a limit on the number of `closing_signed` sent without receiving a response)

Reply:
Is the mismatched fee range only possible in this flow (A is funder):
```
A                 B
   --cl_sig-->  (100, 200) fee=150
   <-cl_sig---  (300, 400) fee=350
   <-cl_sig---  (150, 350) fee=175
   --cl_sig-->  (100, 200) fee=175
```

Or is it possible for the fundee B to be on the receiving end of the non-turn-based behavior where A is sending multiple `closing_signed` without a break (I don't think so given my interpretation of the spec)?

I am trying to draft an update to the taproot channels proposal that eliminates signatures in some cases like if the sent fee_range has no overlap (and only if the taproot feature bit was negotiated).

Reply:
I see now that you've answered my question in that A (funder) is the one that sends multiple closing_signed if B sends a `warning`

Reply:
Yes, sorry for the late reply. There are two possible flows for `fee_range` closing. The first is the happy flow:

* A sends `closing_signed`
* B responds with `closing_signed` inside the `fee_range` proposed by A (B is not allowed to answer with a fee outside of A's `fee_range`)
* A responds with `closing_signed` with the same fee as the previous message

The second possible flow is the unhappy flow where B isn't satisfied with A's `fee_range`:

* A sends `closing_signed`
* B doesn't like the `fee_range`, sends a `warning`
* A sends another `closing_signed`
* B still doesn't like it, sends another `warning`
* ...
* A sends another `closing_signed`
* B likes that `fee_range`, responds with `closing_signed`
* A responds with `closing_signed` with the same fee as the previous message

We can make that strictly turn-based by having Bob send a new message (e.g. `reject_closing_signed`) to express his disagreement with the `fee_range`, in which case we could have nonces in both messages.

Reply:
I don't mind doing a coop close v3 for taproot, with a more strictly specified protocol if it's easier than tacking it on to the existing one!

Reply:
Right I dont mind either, I can draft some coop changes to add onto @Roasbeef taproot proposal